Make thermaltime sklearn compatible

SarahAlidoost opened the pull request:
closes https://github.com/phenology/springtime/issues/151

SarahAlidoost committed:
add sklearn wrapper for thermaltimemodel

SarahAlidoost committed:
add test for sklearn thermaltime

SarahAlidoost committed:
add example recipe and python script for thermaltime

SarahAlidoost committed:
add dependencies to setup.py

SarahAlidoost committed:
skip test check_regressors_train due to low score values

SarahAlidoost committed:
add ruamel.yaml and pytest to the setup

SarahAlidoost committed:
run black and flake8

SarahAlidoost committed:
move arguments of fit to init

SarahAlidoost committed:
remove unused files

SarahAlidoost committed:
draft pycaret test

SarahAlidoost committed:
fix TestPyCaretCompliance, add pyphenology data to the test

SarahAlidoost committed:
refactor TestPyCaretCompliance

SarahAlidoost committed:
add some notes, change the test value

SarahAlidoost committed:
remove print, remove exp before load_model

Peter9192 commented on pyPhenology/models/sklearn_thermaltime.py:
```diff
@@ -0,0 +1,77 @@
+    ):
+        # Check that X and y have correct shape
+        X, y = check_X_y(X, y)
+        # raise Exception(X)
```
```suggestion
        
```

Peter9192 commented on pyPhenology/models/sklearn_thermaltime.py:
```diff
@@ -0,0 +1,77 @@
+        self.model_ = ThermalTime()
+        predictors = pd.DataFrame(
+            {
+                "year": len(X.flatten()) * [2000],
```
This seems like a very particular way to derive a list of years from incoming data. Can you clarify how/why this works?

Peter9192 replied:
Also could you get `len(X.flatten())` by accessing `X.size` instead?

SarahAlidoost replied:
More comments are added. Also, `len(X.flatten())` is replaced with `X.size`

Peter9192 commented on pyPhenology/models/sklearn_thermaltime.py:
```diff
@@ -0,0 +1,77 @@
+        predictors = pd.DataFrame(
+            {
+                "year": len(X.flatten()) * [2000],
+                "site_id": np.repeat(np.arange(len(X)), X.shape[1]),
```
Do I understand correctly that your pretending every row is a unique site as a workaround for the expected structure by pyphenology?

SarahAlidoost replied:
More comments are added.

Peter9192 commented on pyPhenology/models/sklearn_thermaltime.py:
```diff
@@ -0,0 +1,77 @@
+            {
+                "year": len(X.flatten()) * [2000],
+                "site_id": np.repeat(np.arange(len(X)), X.shape[1]),
+                "doy": list(range(X.shape[1])) * X.shape[0],
```
So here you're assuming that all columns of the incoming data correspondond to a DOY, starting at one and ending at `n`? (where `n` is the number of columns)? Would be good to document that this is the expected structure for this wrapper.

SarahAlidoost replied:
More comments are added

Peter9192 commented on pyPhenology/models/sklearn_thermaltime.py:
```diff
@@ -0,0 +1,77 @@
+
+        # Define temperature, doy_series
+        self.model_ = ThermalTime()
+        predictors = pd.DataFrame(
```
```suggestion
        # Convert incoming data to expected structure as documented here
        # https://pyphenology.readthedocs.io/en/master/data_structures.html
        predictors = pd.DataFrame(
```

Peter9192 commented on test/test_sklearn_thermaltime.py:
```diff
@@ -0,0 +1,87 @@
+        rmse_phenology = np.sqrt(np.mean((predicted_doy - observed_doy) ** 2))
+
+        # Prepare data to train the model using pycaret
+        doy_array, temperature_array, _ = mu.misc.temperature_only_data_prep(
+            observations, predictors, for_prediction=False
+        )
```
So here we're using pyphenology's internals to tranform the data to the "sklearn" format. Presumably pyphenology also uses this transformation internally. So that means, in our wrapper, we're going from sklearn compatible to pyphenology compatible, and then inside pyphenology we go back to sklearn compatible. 

I wonder if it isn't possible to hook into the existing pyphenology code only after this transformation - that would save us having to transform back and forth between these two formats alltogether.

SarahAlidoost replied:
yes, good point. I submitted issue #3

Peter9192 reviewed (changes requested):
Nice work @SarahAlidoost, I like how you test by comparing "raw" vs "wrapped" pyphenology. It's also a quite elegant and simple wrapper - we could consider shipping it on it's own and simply importing pyphenology. 

In either case,  a bit of explanation of how to use this wrapper on top of the readme or so would be very helpful. 

My main question though is about the transformations back and forth, and generalization to other models. It looks like were going through a bit of a hassle to get to the data format pyphenology exposes, but then internally pyphenology more or less tranforms back to where we were coming from. Sounds like with a bit of effort, we could eliminate all these transforms alltogether - though we might have to come up with a standardized column naming scheme. 

My other question is about generalizing this to other pyphenology models. Opened #1 to discuss that further.

SarahAlidoost committed:
set seed in pyphenology, set preprocess false in pycaret

SarahAlidoost committed:
remove a comment

Co-authored-by: Peter Kalverla <peter@kalversweep.nl>

SarahAlidoost committed:
add a comment

Co-authored-by: Peter Kalverla <peter@kalversweep.nl>

SarahAlidoost committed:
replace len with size

SarahAlidoost committed:
add more comments about converting incoming data to pyphenology structure

SarahAlidoost commented:
> Nice work @SarahAlidoost, I like how you test by comparing "raw" vs "wrapped" pyphenology. It's also a quite elegant and simple wrapper - we could consider shipping it on it's own and simply importing pyphenology.
> 
> In either case, a bit of explanation of how to use this wrapper on top of the readme or so would be very helpful.
> 
> My main question though is about the transformations back and forth, and generalization to other models. It looks like were going through a bit of a hassle to get to the data format pyphenology exposes, but then internally pyphenology more or less tranforms back to where we were coming from. Sounds like with a bit of effort, we could eliminate all these transforms alltogether - though we might have to come up with a standardized column naming scheme.
> 
> My other question is about generalizing this to other pyphenology models. Opened #1 to discuss that further.

thanks for the helpful review. I addressed your comments. Please have another look. Meantime, I found how to set `seed` in pyPhenology, see my [commit](https://github.com/phenology/pyPhenology/pull/1/commits/06cb19a15a7d901e7d1153763bdc0325b8663c89).

Peter9192 reviewed (approved):
Thanks for addressing the comments! Let's continue the great work in #2 and or #3 👍
